_posts/2023-06-04-not-generating-constructors.md

```diff
@@ -0,0 +1,411 @@
+---
+tags:
+- Subspace
+- C++
+- Performance
+- Memory Safety
+---
+
+# Avoiding Death by a Thousand Constructors
+
+**tl;dr We are going to see how Subspace types allow a zero-cost transition from native pointers to
+bounded view types (which means we can get spatial memory safety) that can't be achieved with the
+standard library types.**
+
+Long ago, C and C++ programmers passed around pointers to arrays, hopefully with a size to indicate
+how large the array was in order avoid reading/writing out of bounds, though frequently not. This is
+still a common pattern, though "modern C++" has provided tools to eliminate this class of memory
+safety bug, though the use of "view" types. These are types like `std::string_view` or `std::span`
+in the standard library. Unfortunately, these types don't actually check for using memory out of
+bounds in the C++ standard.
+
+Much has been written about the memory unsafety of lack of bounds checking, and we (as in the
+software industry) have known ways to solve this simple issue for decades now.
+* Here's an article from 1996 (almost 30 years ago)
+[proposing a bounds-checked structure](https://ecs.syr.edu/faculty/fawcett/Handouts/cse775/presentations/BruceMcKinneyPapers/safeArrays.htm)
+for interacting with
+[OLE](https://learn.microsoft.com/en-us/cpp/mfc/ole-background).
+* Here's a [NIST paper titled "C++ in Safety Critical Systems"](
+https://upload.wikimedia.org/wikipedia/commons/5/50/C%2B%2B_in_safety_critical_systems_%28IA_cinsafetycritica5769bink%29.pdf)
+from 1995 which states "Class libraries that provide smart (safe) pointers and array bounds
+checking should be used."
+
+In 2022, The Clang/Libc++ project accepted the
+[C++ Buffer Hardening](https://discourse.llvm.org/t/rfc-c-buffer-hardening/65734) proposal that
+makes it possible to enable hardening checks inside these standard library types with bounds.
+Though it requires you to build your own standard library. The one that shipped with your operating
+system likely [doesn't have this enabled](https://godbolt.org/z/Y4n99dx5j). At least your
+[web browser](https://bugs.chromium.org/p/chromium/issues/detail?id=1335422) may have this enabled
+(Chrome has since Q1 2023, but I don't know about any other Browser,
+[Chromium-derived or not](https://chromium.googlesource.com/chromium/src/+/master/docs/security/faq.md#how-can-i-know-which-fixes-to-include-in-my-downstream-project)).
+
+The same proposal introduces the [-Wunsafe-buffer-usage](

```

Comment:
If GitHub's flavor of Markdown allows it (Captain Kirk screaming Khaaaaaaaan.gif), I'd use code formattin' for `-Wunsafe-buffer-usage`.